### ISSUE TITLE
Contribute examples to Rosetta Code

### ISSUE THREAD

**Issue by mlhaufe** [carp-lang/Carp#222]
Contributing examples to Rosetta Code would be useful and informative not only for newcomers to the languages but as a sanity check on the syntax/semantics of the languages

http://rosettacode.org/wiki/Category:Programming_Languages

**Comment by hellerve** [comment]
While I agree with this, I think the standard library and compiler are still very much in flux. This makes it hard to program “good” solutions that will stay idiomatic for some time.

While it would surely be great from a PR perspective, I’m not sure that programming and maintaining the example wouldn’t—at the moment—be more of a distraction than something that adds value.

Maybe I’m wrong though. I’m bad at juding what’s important and what isn’t.

**Comment by eriksvedang** [comment]
Let's do this when we've reached some kind of milestone, after doing some more hard thinking about naming and such!

**Comment by mlhaufe** [comment]
> I’m bad at juding what’s important and what isn’t.

My only feedback to this would be to solve the Expression Problem. That would impact language semantics the greatest.

**Comment by eriksvedang** [comment]
@mlhaufe I believe we have a solution for it already, namely interfaces. Or do you disagree? :) I'm more concerned with the current lack of lambdas, I think a lot of code will change when we have those.

**Comment by mlhaufe** [comment]
@eriksvedang Indeed, I disagree. This topic should be a separate issue if you with to pursue it.

Basically here's what you have to solve:

(Examples in TypeScript)
https://gist.github.com/mlhaufe/0f40e8a32a9e9e4e91c501c7656a79f8

The functional form in this case.

**Comment by eriksvedang** [comment]
Here's how I'd solve that in Carp: https://github.com/carp-lang/Carp/blob/master/examples/expression_problem.carp

**Comment by mlhaufe** [comment]
Can you represent Hutton’s Razor in your example explicitly? 

Create a type representing the AST of expressions: 

`data Exp = Lit Int | Add Int Int`

evaluate the datatype:

```
eval :: Exp => Int
eval(Lit n) = n
eval(Add x y) = (eval x) + (eval y)
```

Now extend Exp data type with "Mul Int Int" without modifying the above code
Now extend eval to support the new expression without modifying the above code

**Comment by hellerve** [comment]
We don’t have sum types yet, but we can do the following:
```
(deftype Lit [n Int])
(deftype Add [x Int, y Int]) 

(definterface eval (Fn [a] Int))

(defmodule Lit
   (defn eval [lit]
      @(n lit))
)

(defmodule Add
   (defn eval [add]
      (+ @(x add) @(y add)))
)

(defn main []
   (+ (eval &(Lit.init 1)) (eval &(Add.init 3 4))))
```

We could extend that like so:

```
(deftype Mul [x Int, y Int])

(defmodule Mul
   (def eval [mul]
      (* @(x mul) @(y mul)))
)
```

See also [how protocols solve the expression problem](https://blog.rapid7.com/2016/12/07/solving-the-expression-problem/). Interfaces are just our flavour of it.

**Comment by mlhaufe** [comment]
Indeed, I am aware of the application of protocols but the issue here is deeper. You might be interested in the LTU conversation on this and Clojure's attempt at it:

http://lambda-the-ultimate.org/node/4136#comment-63045

Again, *if* this is an interest, I'd think it should be a separate thread.

**Comment by hellerve** [comment]
The thread you linked to says what Clojure is missing is static typing. We have static typing, and we have solved the problems you gave us.

Do you have any other objections?

**Comment by mlhaufe** [comment]
Correct my if I'm wrong, but there is an implied tone here. I'll wait to see how the language develops further before considering further feedback. 

I do have other issues, but I doubt it matters at this stage of the language's development as it is so much in flux as mentioned. My primary feedback was always the Rosetta Code suggestion and not the Expression Problem. I can very much push that door further to point out more shortcomings but what's the point?

**Comment by eriksvedang** [comment]
Yes, agreed. I'll leave this issue open as a reminder to submit to Rosetta when we've reached a more stable point.

**Comment by hellerve** [comment]
@mlhaufe Pardon the tone. This is not how I want to interact with you—or anyone else for that matter.

I would be interested in hearing more about your complaints with regards to the expression problem, either on another issue or using another channel altogether. I don’t want to derail the point of this issue even more, but I believe that I owed you an apology for the tone of the reply above.
